lib/commands/help.js

```diff
@@ -103,36 +87,30 @@ module.exports = Command.extend({
       // Iterate through each arg beyond the initial 'help' command,
       // and try to display usage instructions.
       rawArgs.forEach(function(commandName) {
-        this._printDetailedHelpForCommand(commandName, commandOptions, json);
+        this._printHelpForCommand(commandName, true, commandOptions);
       }, this);
     }
-
-    if (commandOptions.json) {
-      this._printJsonHelp(json);
-    }
   },
 
-  _printBasicHelpForCommand: function(commandName, options, json) {
-    if (options.json) {
-      this._addCommandHelpToJson(commandName, false, options, json);
-    } else {
-      this._printHelpForCommand(commandName, false, options);
-    }
-  },
+  _printJsonHelp: function(commandOptions, rawArgs) {
+    var generator = new JsonGenerator({
+      ui: this.ui,
+      project: this.project,
+      commands: this.commands,
+      tasks: this.tasks
+    });
 
-  _printDetailedHelpForCommand: function(commandName, options, json) {
-    if (options.json) {
-      this._addCommandHelpToJson(commandName, true, options, json);
-    } else {
-      this._printHelpForCommand(commandName, true, options);
-    }
-  },
+    var json = generator.generate(commandOptions, rawArgs);
 
-  _addCommandHelpToJson: function(commandName, single, options, json) {
-    var command = this._lookupCommand(commandName);
-    if (!command.skipHelp || single) {
-      json.commands.push(command.getJson(options));
-    }
+    var outputJsonString = JSON.stringify(json, function(key, value) {
+      // build command has a recursive property
```

Comment:
I think we should fix that instead of introducing a mapping function.
the [build command](https://github.com/Turbo87/ember-cli/blob/json/lib/commands/build.js#L14) should specify `type: String` as it is done [here](https://github.com/Turbo87/ember-cli/blob/json/lib/commands/test.js#L31).
If you do so, you can get rid of [this line](https://github.com/Turbo87/ember-cli/blob/json/lib/commands/build.js#L3) aswell. same would need to be done [here](https://github.com/Turbo87/ember-cli/blob/json/lib/commands/serve.js#L33).

Then you could rewrite this hole thing into
```javascript
var outputJsonString = JSON.stringify(json, null, 2);
```